src/handler.ts

```diff
@@ -87,36 +93,67 @@ export class DebuggerHandler {
     widget: DebuggerHandler.SessionWidget[DebuggerHandler.SessionType],
     connection: Session.ISessionConnection
   ): Promise<void> {
-    const updateHandler = async () => {
+    if (!connection) {
+      delete this._kernelChangedHandlers[widget.id];
+      delete this._statusChangedHandlers[widget.id];
       return this._update(widget, connection);
+    }
+
+    const kernelChanged = () => {
+      void this._update(widget, connection);
     };
+    const kernelChangedHandler = this._kernelChangedHandlers[widget.id];
 
-    // setup handler when the kernel changes
-    const kernelChangedHandler = this._kernelChangedHandlers[connection.path];
     if (kernelChangedHandler) {
       connection.kernelChanged.disconnect(kernelChangedHandler);
     }
-    connection.kernelChanged.connect(updateHandler);
-    this._kernelChangedHandlers[connection.path] = updateHandler;
+    this._kernelChangedHandlers[widget.id] = kernelChanged;
+    connection.kernelChanged.connect(kernelChanged);
```

Comment:
Do we need to keep track of both `kernelChanged` signals?

The one from `SessionConnection` should be proxied by the `SessionContext`: https://github.com/jupyterlab/jupyterlab/blob/d3cc2470976d3ea285abc6f855ad193445fca9e7/packages/apputils/src/sessioncontext.tsx#L712

Reply:
The answer is most likely yes because of the files.